standardization/src/main/scala/za/co/absa/enceladus/standardization/interpreter/StandardizationInterpreter.scala

```diff
@@ -17,109 +17,129 @@ package za.co.absa.enceladus.standardization.interpreter
     if (validationErrors._1.nonEmpty) {
       throw new ValidationException("A fatal schema validation error occurred.", validationErrors._1)
     }
+  }
 
-    // TODO: remove when spark-xml handles empty arrays
-    val df1: Dataset[Row] = if (inputType.toLowerCase() == "xml") {
-      df.select(expSchema.fields.map { field: StructField =>
-        SparkXMLHack.hack(field, "", df).as(field.name)
-      }: _*)
-    } else df
+  private def standardizeTheDataset(df: Dataset[Row], expSchema: StructType)
+                                   (implicit spark: SparkSession, udfLib: UDFLibrary): (DataFrame, ErrorCols)  = {
 
-    logger.info(s"Step 2: Standardization")
-    // step 2 - standardize
-    val std: DataFrame = df1.select(expSchema.fields.flatMap { field: StructField =>
+    val(cols, errorColNames, _) = expSchema.fields.foldLeft(List.empty[Column], List.empty[String], 1L) { (acc, field) =>
       logger.info(s"Standardizing field: ${field.name}")
+      val (accCols, accErrorCols, nextErrorColumnIndex) = acc
       if (field.name == ErrorMessage.errorColumnName) {
-        Seq(df1.col(field.name))
+        (df.col(field.name)::accCols, accErrorCols, nextErrorColumnIndex)
       } else {
         val stdOutput = TypeParser.standardize(field, "", df.schema)
         logger.info(s"Applying standardization plan for ${field.name}")
-        stdApplyLogic(stdOutput, field)
+        val (column1, column2, errField) = stdApplyLogic(stdOutput, field, nextErrorColumnIndex)
+        (column1::column2::accCols, errField::accErrorCols, nextErrorColumnIndex + 1)
       }
-    }: _*)
+    }
 
-    // step 3 drop intermediate error columns
-    logger.info(s"Step 3.1: Preserve existing error column")
-    // preserve existing errors - rename into a temporary column
-    val errColPreserved = if (std.columns.contains(ErrorMessage.errorColumnName)) {
-      val newName = ErrorMessage.errorColumnName + (Random.nextInt.abs)
-      errorCols.append(newName)
-      std.withColumnRenamed(ErrorMessage.errorColumnName, newName)
-    } else std
+    (df.select(cols.reverse: _*), errorColNames)
+  }
 
-    logger.info(s"Step 3.2: Collect all temporary error columns")
+  private def preserveExistingErrorColumn(dataFrame: DataFrame, errorCols: ErrorCols): (DataFrame, ErrorCols) = {
+    if (dataFrame.columns.contains(ErrorMessage.errorColumnName)) {
+      // preserve existing errors - rename into a temporary column
+      val newName = s"${ErrorMessage.errorColumnName}0"
+      (dataFrame.withColumnRenamed(ErrorMessage.errorColumnName, newName), newName::errorCols)
+    } else {
+      (dataFrame, errorCols)
+    }
+  }
+
+  private def collectAllTempErrorColumns(dataFrame: DataFrame, errorCols: ErrorCols)
+                                            (implicit spark: SparkSession, udfLib: UDFLibrary): DataFrame = {
     // collect all of the error attributes into an array and flatten
     logger.info(s"Error cols: ${errorCols.toList}")
```

Comment:
```suggestion
     logger.info(s"Error cols: $errorCols")
```

It already is a list.